services/autoscaling/src/simcore_service_autoscaling/modules/dask.py

```diff
@@ -0,0 +1,212 @@
+    except OSError as exc:
+        raise DaskSchedulerNotFoundError(url=url) from exc
+
+
+async def list_unrunnable_tasks(url: AnyUrl) -> list[DaskTask]:
+    """
+    Raises:
+        DaskSchedulerNotFoundError
+    """
+
+    def _list_tasks(
+        dask_scheduler: distributed.Scheduler,
+    ) -> dict[str, dict[str, Any]]:
+        return {
+            task.key: task.resource_restrictions for task in dask_scheduler.unrunnable
+        }
+
+    async with _scheduler_client(url) as client:
+        list_of_tasks: dict[
+            DaskTaskId, DaskTaskResources
+        ] = await _wrap_client_async_routine(client.run_on_scheduler(_list_tasks))
+        _logger.info("found unrunnable tasks: %s", list_of_tasks)
+        return [
+            DaskTask(task_id=task_id, required_resources=task_resources)
+            for task_id, task_resources in list_of_tasks.items()
+        ]
+
+
+async def list_processing_tasks(url: AnyUrl) -> list[DaskTaskId]:
+    """
+    Raises:
+        DaskSchedulerNotFoundError
+    """
+    async with _scheduler_client(url) as client:
+        processing_tasks = set()
+        if worker_to_processing_tasks := await _wrap_client_async_routine(
+            client.processing()
+        ):
+            _logger.info("cluster worker processing: %s", worker_to_processing_tasks)
+            for tasks in worker_to_processing_tasks.values():
+                processing_tasks |= set(tasks)
+
+        return list(processing_tasks)
+
+
+DaskWorkerUrl = str
+DaskWorkerDetails = dict[str, Any]
+
+
+def _dask_worker_from_ec2_instance(
+    client: distributed.Client, ec2_instance: EC2InstanceData
+) -> tuple[DaskWorkerUrl, DaskWorkerDetails]:
+    """
+    Raises:
+        Ec2InvalidDnsNameError
+        DaskNoWorkersError
+        DaskWorkerNotFoundError
+    """
+    node_hostname = node_host_name_from_ec2_private_dns(ec2_instance)
+    node_ip = node_ip_from_ec2_private_dns(ec2_instance)
+    scheduler_info = client.scheduler_info()
+    assert client.scheduler  # nosec
+    if "workers" not in scheduler_info or not scheduler_info["workers"]:
+        raise DaskNoWorkersError(url=client.scheduler.address)
+    workers: dict[DaskWorkerUrl, DaskWorkerDetails] = scheduler_info["workers"]
+
+    _logger.debug("looking for %s in %s", f"{ec2_instance=}", f"{workers=}")
+
+    # dict is of type dask_worker_address: worker_details
+    def _find_by_worker_host(
+        dask_worker: tuple[DaskWorkerUrl, DaskWorkerDetails]
+    ) -> bool:
+        _, details = dask_worker
+        return bool(details["host"] == node_ip) or bool(
+            node_hostname in details["name"]
+        )
+
+    filtered_workers = dict(filter(_find_by_worker_host, workers.items()))
+    if not filtered_workers:
+        raise DaskWorkerNotFoundError(
+            worker_host=ec2_instance.aws_private_dns, url=client.scheduler.address
+        )
+    assert len(filtered_workers) == 1  # nosec
+    return next(iter(filtered_workers.items()))
```

Comment:
python has such an ugly syntax for these cases!

Reply:
yep